zmon-security-common/src/main/java/de/zalando/zmon/security/tvtoken/TvTokenService.java

```diff
@@ -0,0 +1,41 @@
+package de.zalando.zmon.security.tvtoken;
+
+import org.springframework.util.Base64Utils;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+
+import de.zalando.zmon.persistence.OnetimeTokensSProcService;
+
+public class TvTokenService {
+
+    private final OnetimeTokensSProcService oneTimeTokenSProcService;
+
+    private final Joiner cookieValueJoiner = Joiner.on("|").useForNull("UNKNOWN");
+    private final Splitter cookieValueSplitter = Splitter.on("|").omitEmptyStrings();
+
+    public TvTokenService(OnetimeTokensSProcService onetimeTokensSProcService) {
+        this.oneTimeTokenSProcService = onetimeTokensSProcService;
+    }
+
+    public String createCookieValue(String token, String bindIp, String sessionId) {
+        String cookieValue = cookieValueJoiner.join(token, bindIp, sessionId);
+        String encodedValue = Base64Utils.encodeToString(cookieValue.getBytes());
+        return encodedValue;
+    }
+
+    public Iterable<String> decodeCookieValue(String encodedCookieValue) {
+        String decodedValue = new String(Base64Utils.decodeFromString(encodedCookieValue));
+        return cookieValueSplitter.split(decodedValue);
+    }
+
+    // TODO use sproc
+    public boolean isValidToken(String token, String bindIp, String sessionId) {
```

Comment:
It's not there yet, only the happy path is implemented to see that it can work this way. At it's current state when a user opens the page he/she gets the login-page (if no cookie 'ZMON_TV' was set before). Then the user uses '/tv/123456789' to use the token. When successfully bound with sproc (not implemented yet, always returns 'true' at the moment) we create a cookie, set a value (whatever should be in there) and add that cookie to the response with a redirect to '/' to retrigger authentication again through filterchain. But now the the Remember-Me service kicks in because no Authentication-Provider could authenticate (as before) and is invoked as last element in the authentication-chain. Can verify ip and cookie-value and if everything fine returns an 'RememberMeAuthenticationToken'-object and the dashboard is presented.

You can try it yourself, but you should use an 'fresh' browser, then the automatic login with github will stop githubs login-page. Then provide an token via '/tv/1237636485' and you should be logged in.

So, this way it could work. Please define what has to go into the cookie-value. IMO it is also not necessary to store the JSESSIONID in the cookie. We do not depend on any specific session. The JSESSIONID is also not needed in the DB I would say. But we can discuss that. We only have to check 'token'  <-> 'ip' should be same as on bind-time (first time used).